plugins/common/adx/adx.go

```diff
@@ -0,0 +1,206 @@
+)
+
+type Config struct {
+	Endpoint        string          `toml:"endpoint_url"`
+	Database        string          `toml:"database"`
+	Timeout         config.Duration `toml:"timeout"`
+	MetricsGrouping string          `toml:"metrics_grouping_type"`
+	TableName       string          `toml:"table_name"`
+	CreateTables    bool            `toml:"create_tables"`
+	IngestionType   string          `toml:"ingestion_type"`
+}
+
+type Client struct {
+	cfg       *Config
+	conn      *kusto.ConnectionStringBuilder
+	client    *kusto.Client
+	ingestors map[string]ingest.Ingestor
+	logger    telegraf.Logger
+}
+
+func (cfg *Config) NewClient(app string, log telegraf.Logger) (*Client, error) {
+	if cfg.Endpoint == "" {
+		return nil, errors.New("endpoint configuration cannot be empty")
+	}
+	if cfg.Database == "" {
+		return nil, errors.New("database configuration cannot be empty")
+	}
+
+	cfg.MetricsGrouping = strings.ToLower(cfg.MetricsGrouping)
+	if cfg.MetricsGrouping == SingleTable && cfg.TableName == "" {
+		return nil, errors.New("table name cannot be empty for SingleTable metrics grouping type")
+	}
+
+	if cfg.MetricsGrouping == "" {
+		cfg.MetricsGrouping = TablePerMetric
+	}
+
+	if !(cfg.MetricsGrouping == SingleTable || cfg.MetricsGrouping == TablePerMetric) {
+		return nil, errors.New("metrics grouping type is not valid")
+	}
+
+	if cfg.Timeout == 0 {
+		cfg.Timeout = config.Duration(20 * time.Second)
+	}
+
+	if cfg.IngestionType == "" {
+		cfg.IngestionType = QueuedIngestion
+	} else if !(choice.Contains(cfg.IngestionType, []string{ManagedIngestion, QueuedIngestion})) {
+		return nil, fmt.Errorf("unknown ingestion type %q", cfg.IngestionType)
+	}
+
+	conn := kusto.NewConnectionStringBuilder(cfg.Endpoint).WithDefaultAzureCredential()
+	conn.SetConnectorDetails("Telegraf", internal.ProductToken(), app, "", false, "")
+	client, err := kusto.New(conn)
+	if err != nil {
+		return nil, err
+	}
+	return &Client{
+		cfg:       cfg,
+		conn:      conn,
+		ingestors: make(map[string]ingest.Ingestor),
+		logger:    log,
+		client:    client,
+	}, nil
+}
+
+const (
+	TablePerMetric = "tablepermetric"
+	SingleTable    = "singletable"
+	// These control the amount of memory we use when ingesting blobs
+	bufferSize       = 1 << 20 // 1 MiB
+	maxBuffers       = 5
+	ManagedIngestion = "managed"
+	QueuedIngestion  = "queued"
+)
+
+// Clean up and close the ingestor
+func (adx *Client) Close() error {
+	var errs []error
+	for _, v := range adx.ingestors {
+		if err := v.Close(); err != nil {
+			// accumulate errors while closing ingestors
+			errs = append(errs, err)
+		}
+	}
+	if err := adx.client.Close(); err != nil {
+		errs = append(errs, err)
+	}
+
+	adx.client = nil
+	adx.ingestors = nil
+
+	if len(errs) == 0 {
+		return nil
+	}
+	// Combine errors into a single object and return the combined error
+	return kustoerrors.GetCombinedError(errs...)
+}
+
+func (adx *Client) PushMetrics(format ingest.FileOption, tableName string, metricsArray []byte) error {
```

Comment:
Maybe
```suggestion
func (adx *Client) PushMetrics(format ingest.FileOption, tableName string, metrics []byte) error {
```
for brevity?